File: mpi/exercises/exercise_6/provided.c
#include <stdio.h>
#include <mpi.h>

/**
 * @brief Exercise to replace blocking with non-blocking communications.
 * @details This application consists of 2 MPI processes, a sender and a
 * receiver. They exchange an integer using a blocking send and a blocking
 * receive. The aim is to replace these with their non-blocking counterparts and
 * handle the non-blocking communication.
 **/
int main(int argc, char* argv[])
{
	MPI_Init(&argc, &argv);

	int comm_size;
	MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
	if(comm_size != 2)
	{
		printf("This application must be run with 2 MPI processes.\n");
		MPI_Abort(MPI_COMM_WORLD, -1);
	}

	int my_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
	if(my_rank == 0)
	{
		int value_sent = 12345;
		printf("[MPI process %d] I sent value %d.\n", my_rank, value_sent);
		MPI_Send(&value_sent, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
	}
	else
	{
		int value_received = 0;
		MPI_Recv(&value_received, 1, MPI_INT, 0, MPI_ANY_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
		printf("[MPI process %d] I received value %d.\n", my_rank, value_received);
	}

	MPI_Finalize();

	return 0;
}

File: mpi/exercises/exercise_6/provided.f90
!> @brief Exercise to replace blocking with non-blocking communications.
!> @details This application consists of 2 MPI processes, a sender and a
!> receiver. They exchange an integer using a blocking send and a blocking
!> receive. The aim is to replace these with their non-blocking counterparts and
!> handle the non-blocking communication.
PROGRAM main
    USE mpi

    IMPLICIT NONE

    INTEGER :: ierror
    INTEGER :: comm_size
    INTEGER :: my_rank
    INTEGER :: value

    CALL MPI_Init(ierror)

    CALL MPI_Comm_size(MPI_COMM_WORLD, comm_size, ierror)
    IF (comm_size .NE. 2) THEN
        WRITE(*, '(A)') 'This application must be run with 2 MPI processes.'
        CALL MPI_Abort(MPI_COMM_WORLD, -1, ierror)
    END IF

    CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank, ierror)
    IF (my_rank == 0) THEN
        value = 12345
        WRITE(*, '(A,I0,A,I0,A)') '[MPI process ', my_rank, '] I sent value ', value, '.'
        CALL MPI_Send(value, 1, MPI_INTEGER, 1, 0, MPI_COMM_WORLD, ierror)
    ELSE
        value = 0
        CALL MPI_Recv(value, 1, MPI_INTEGER, 0, MPI_ANY_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror)
        WRITE(*, '(A,I0,A,I0,A)') '[MPI process ', my_rank, '] I received value ', value, '.'
    END IF

    CALL MPI_Finalize(ierror)
END PROGRAM main